ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/CastInterpreter.java

```diff
@@ -0,0 +1,253 @@
+      final CastFunction itemCastFunction = castFunction(
+          fromArray.getItemType(), toArray.getItemType(), config);
+      return o ->
+          CastEvaluator.castArray((List<?>) o, itemCastFunction::cast);
+    }
+    throw new KsqlException(getErrorMessage(ConversionType.CAST, from, to));
+  }
+
+  public static CastFunction castToMapFunction(
+      final SqlType from,
+      final SqlType to,
+      final KsqlConfig config
+  ) {
+    if (from.baseType() == SqlBaseType.MAP
+        && to.baseType() == SqlBaseType.MAP) {
+      final SqlMap fromMap = (SqlMap) from;
+      final SqlMap toMap = (SqlMap) to;
+      final CastFunction keyCastFunction = castFunction(
+          fromMap.getKeyType(), toMap.getKeyType(), config);
+      final CastFunction valueCastFunction = castFunction(
+          fromMap.getValueType(), toMap.getValueType(), config);
+      return o ->
+          CastEvaluator.castMap((Map<?, ?>) o, keyCastFunction::cast, valueCastFunction::cast);
+    }
+    throw new KsqlException("Unsupported cast to " + to + ": " + from);
+  }
+
+  /**
+   * Conversion functions that work for converting type either during comparison or casting.
+   */
+  public static class NumberConversions {
+    public static Double toDouble(final Object object, final SqlType from,
+        final ConversionType type) {
+      if (object == null) {
+        return null;
+      }
+      if (object instanceof Number) {
+        return ((Number) object).doubleValue();
+      } else {
+        throw new KsqlException(getErrorMessage(type, from, SqlTypes.DOUBLE));
+      }
+    }
+
+    public static Long toLong(final Object object, final SqlType from, final ConversionType type) {
+      if (object == null) {
+        return null;
+      }
+      if (object instanceof Number) {
+        return ((Number) object).longValue();
+      } else {
+        throw new KsqlException(getErrorMessage(type, from, SqlTypes.BIGINT));
+      }
+    }
+
+    public static Integer toInteger(final Object object, final SqlType from,
+        final ConversionType type) {
+      if (object == null) {
+        return null;
+      }
+      if (object instanceof Number) {
+        return ((Number) object).intValue();
+      } else {
+        throw new KsqlException(getErrorMessage(type, from, SqlTypes.INTEGER));
+      }
+    }
+  }
+
+  private static String getErrorMessage(
+      final ConversionType type,
+      final SqlType from,
+      final SqlType to
+  ) {
+    switch (type) {
+      case CAST:
+        return String.format("Unsupported cast from %s to %s", from, to);
+      case COMPARISON:
+        return String.format("Unsupported comparison between %s and %s", from, to);
+      case ARITHMETIC:
+        return String.format("Unsupported arithmetic between %s and %s", from, to);
+      default:
+        throw new KsqlException("Unknown Conversion type " + type);
+    }
+  }
+
+  public enum ConversionType {
+    CAST,
+    COMPARISON,
```

Comment:
Do we use `COMPARISON` and `ARITHMETIC` today?

Reply:
Yeah, it was called in other places depending on the context just to give a slightly more appropriate error message, but this doesn't add much.  I'm removing this enum.